src/main/resources/xsl/openagrar-solr.xsl

```diff
@@ -60,19 +59,41 @@
         </field>
       </xsl:if>
     </xsl:for-each>
-    <xsl:for-each
+
+	<xsl:for-each
       select="mods:name[mods:role/mods:roleTerm[@authority='marcrelator' and (@type='text' and text()='author') or (@type='code' and text()='aut')]]">
-      <xsl:if test="position()= last()">
-        <field name="mods.lastAuthor">
-          <xsl:for-each select="mods:displayForm | mods:namePart | text()">
-            <xsl:value-of select="concat(' ',mcrxsl:normalizeUnicode(.))" />
-          </xsl:for-each>
-        </field>
-        <field name="mods.lastAuthor.affiliation">
-          <xsl:value-of select="mods:affiliation" />
-        </field>
-      </xsl:if>
-    </xsl:for-each>
+		  <xsl:variable name="autrole">
+			  <xsl:choose>
+				<xsl:when test="./mods:role/mods:roleTerm[substring-after(@valueURI, '#')='main_author']">
+					<xsl:text>mainAuthor</xsl:text>
+				</xsl:when>
+				<xsl:when test="count(//mods:roleTerm[substring-after(@valueURI, '#')='main_author'])=0 and position()=1">
+					<xsl:text>mainAuthor</xsl:text>
+				</xsl:when>
+				<xsl:when test="./mods:role/mods:roleTerm[substring-after(@valueURI, '#')='co_author']">
+					<xsl:text>coAuthor</xsl:text>
+				</xsl:when>
+				<xsl:when test="./mods:role/mods:roleTerm[substring-after(@valueURI, '#')='corresponding_author']">
+					<xsl:text>correspondingAuthor</xsl:text>
+				</xsl:when>
+				<xsl:when test="./mods:role/mods:roleTerm[substring-after(@valueURI, '#')='last_author']">
+					<xsl:text>lastAuthor</xsl:text>
+				</xsl:when>
+				<xsl:when test="count(//mods:roleTerm[substring-after(@valueURI, '#')='last_author'])=0 and position()=last()">
```

Comment:
Bitte die Bedingung
//mods:roleTerm[substring-after(@valueURI, '#')='last_author'])=0
nach 
//mods:roleTerm[substring-after(@valueURI, '#'))=0
ändern.

Damit wird verhindert, dass beide Varianten gleichzeitig genutzt werden.

Reply:
Beispiel 3: Ein Corresponding Author und sonnst nur Co Autoren:
https://reposis-test.gbv.de/openagrar/receive/zenodo_mods_00000080?XSL.Style=xml
https://reposis-test.gbv.de/openagrar/receive/zenodo_mods_00000080?XSL.Style=solrdocument
Aber obwohl es keinen Main Author gibt, erscheint einer im Solrdocument. Und die Person ist sogar als Co Autor ausgezeichnet.

Beispiel 4: Ein Main Author und kein Last Author.
https://reposis-test.gbv.de/openagrar/receive/openagrar_mods_00083964?XSL.Style=xml
https://reposis-test.gbv.de/openagrar/receive/openagrar_mods_00083964?XSL.Style=solrdocument
Aber obwohl es keinen Last Author gibt, erscheint einer im Solrdocument.

Reply:
Lange Rede kurzer Sinn: Ich hätte Code in der Form erwartet:

      <xsl:choose>
        <xsl:when test="../mods:name/mods:roleTerm[substring-after(@valueURI, '#'))=0">
          <!-- Neuer Code -->
        </xsl:when>
        <xsl:otherwise>
          <!-- Alter Code -->
        </xsl:otherwise>
      </xsl:choose>

Reply:
Der Eingabeformular lässt zu, dass einige Autoren mir Rollen versehen sind, andere nicht. In so einem Fall reicht dieser "when"-Test  nicht aus.
Und wie sollen dann Autoren ohne Rolle indexiert werden? Ich dachte: 

- Der letzte Autor als "last Author" fall vorher keine diese Rolle bekam
- Alle anderen als Co_Autor.

Reply:
Wie ich geschrieben habe: Sobald in einem Dokument bei einem Autor eine Rolle vergeben wurde soll der alte Mechanismus gar nicht mehr greifen. Wenn ein Autor keine Rolle im mods bekommt hat er keine im Solrdocument. Sie sind dann einfach nur Autoren. Das Feld für Autoren gibt es schon.

Die Bearbeiter werden die Autorenrolen auch nur für Autoren aus ihren Instituten vergeben. Und auch nur die Bearbeiter  die die Statistik brauchen. Es wird also sehr viele Autoren ohne Rolle geben.

Die alte Variante soll zukünftig total wegfallen, wenn die Autorenrollen alle für die Statstik relevanten Dokumente nachgetragen sind. Dies werde ich jedoch noch mit den OpenAgrarlern absprechen.